### PR TITLE
[IMPROVE] Add support to range downloads on file system storage

### PR DESCRIPTION
**Pull request by sampaiodiego**
## Issue(s)

Closes #20203
Related to https://github.com/RocketChat/Rocket.Chat.ReactNative/issues/2759

### CODE REVIEW

File: app/file-upload/server/lib/ranges.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
+export function getByteRange(header) {
+	if (!header) {
+		return;
+	}
+	const matches = header.match(/(\d+)-(\d+)/);
+	if (!matches) {
+		return;
+	}
+	return {
+		start: parseInt(matches[1], 10),
```
Comment (KevLehman) [inline app/file-upload/server/lib/ranges.js:10]: This may return `infinity` for veeery big numbers. Is that ok?

Comment (sampaiodiego) [reply app/file-upload/server/lib/ranges.js:10]: This was being used before here: https://github.com/RocketChat/Rocket.Chat/pull/21463/files#diff-6b22c83cdb33e48bba19cad98bb830a3e373cf4abd6a488f523884a5b2c63cf5L59

so I'm not that worried about it

File: app/file-upload/server/config/FileSystem.js [hunk 2]
```diff
@@ -14,17 +17,32 @@ const FileSystemUploads = new FileUploadClass({
 	get(file, req, res) {
 		const filePath = this.store.getFilePath(file._id, file);
 
+		const options = {};
+
+		if (req.headers.range) {
+			const range = getFileRange(file, req);
+			setRangeHeaders(range, file, res);
+
+			if (range) {
```
[the remark this answers is not in the document]
Comment (sampaiodiego) [reply app/file-upload/server/config/FileSystem.js:26]: right.. I'll change to look more how GridFS does.. 👍

File: app/file-upload/server/config/FileSystem.js [hunk 3]
```diff
@@ -6,6 +6,9 @@ import _ from 'underscore';
 
 import { settings } from '../../../settings';
 import { FileUploadClass, FileUpload } from '../lib/FileUpload';
+import { getFileRange, setRangeHeaders } from '../lib/ranges';
+
+const statSync = Meteor.wrapAsync(fs.stat);
```
Comment (KevLehman) [inline app/file-upload/server/config/FileSystem.js:11]: Isn't a `fs.statSync` function? 👀

File: app/file-upload/server/lib/ranges.js (hunk 1 above)
Comment (KevLehman) [reply app/file-upload/server/lib/ranges.js:10]: Nice! 🥳

File: app/file-upload/server/config/FileSystem.js (hunk 3 above)
Comment (sampaiodiego) [reply app/file-upload/server/config/FileSystem.js:11]: yes. but using `Meteor.wrapAsync` it doesn't make it really sync, it is still doesn't block the event loop, which is better 😬  take a look at https://blog.meteor.com/meteor-fibers-meet-promises-meet-callbacks-a-practical-guide-e134db15afcb

Comment (KevLehman) [reply app/file-upload/server/config/FileSystem.js:11]: Ahhhh! Nice nice, I really thought you wanted to do it "sync" by the name haha.

**Review by KevLehman — APPROVED** [review]

**Pull request merged**